lib/dockerregistry/inventory.go

```diff
@@ -404,6 +395,92 @@ func (sc *SyncContext) getPromotionCandidates(
 	return toPromote
 }
 
+// checkOverlappingEdges checks to ensure that all the edges taken together as a
+// whole are consistent. It checks that there are no duplicate promotions
+// desired to the same destination vertex (same destination PQIN). If the
+// digests are the same for those edges, ignore because by definition the
+// digests are cryptographically guaranteed to be the same thing (it doesn't
+// matter if 2 different parties want the same exact image to become promoted to
+// the same destination --- in many ways this is actually a good thing because
+// it's a form of redundancy). However, return an error if the digests are
+// different, because most likely this is at best just human error and at worst
+// a malicious attack (someone trying to push an image to an endpoint they
+// shouldn't own, or possibly an attempt to rename a tag for an existin image to
+// be something else).
+//
+// nolint[gocyclo]
+func checkOverlappingEdges(
+	edges map[PromotionEdge]interface{}) (map[PromotionEdge]interface{}, error) {
+
+	// Build up a "promotionIntent". This will be checked below.
+	promotionIntent := make(map[string]map[Digest][]PromotionEdge)
+	for edge := range edges {
+		dstPQIN := ToPQIN(edge.DstRegistry.Name,
+			edge.DstImageTag.ImageName,
+			edge.DstImageTag.Tag)
+
+		digestToEdges, ok := promotionIntent[dstPQIN]
+		if ok {
+			// Store the edge in the appropriate bucket.
+			edgeList, ok := digestToEdges[edge.Digest]
```

Comment:
So I think you can do

`digestToEdges[edge.Digest] = append(digestToEdges[edge.Digest], edge)`

https://play.golang.org/p/IZRWwWY7FDG

Reply:
So if I do

```
		digestToEdges, ok := promotionIntent[dstPQIN]
		if ok {
			// Store the edge.
			digestToEdges[edge.Digest] = append(digestToEdges[edge.Digest], edge)
			promotionIntent[dstPQIN] = digestToEdges
		} else {
			// Make this edge lay claim to this destination vertex.
			edgeList := make([]PromotionEdge, 0)
			edgeList = append(edgeList, edge)
			digestToEdges := make(map[Digest][]PromotionEdge)
			digestToEdges[edge.Digest] = edgeList
			promotionIntent[dstPQIN] = digestToEdges
		}
```

it compiles and the unit tests pass. But if I make it slightly more DRY with

```
		digestToEdges, ok := promotionIntent[dstPQIN]
		if ok {
			// Store the edge.
			digestToEdges[edge.Digest] = append(digestToEdges[edge.Digest], edge)
		} else {
			// Make this edge lay claim to this destination vertex.
			edgeList := make([]PromotionEdge, 0)
			edgeList = append(edgeList, edge)
			digestToEdges := make(map[Digest][]PromotionEdge)
			digestToEdges[edge.Digest] = edgeList
		}
		promotionIntent[dstPQIN] = digestToEdges
```

I get this unit test failure:

```
--- FAIL: TestValidateManifestsFromDir (0.01s)
panic: assignment to entry in nil map [recovered]
	panic: assignment to entry in nil map

goroutine 10 [running]:
testing.tRunner.func1(0xc000101200)
	GOROOT/src/testing/testing.go:874 +0x3a3
panic(0x8a0ac0, 0x9975f0)
	GOROOT/src/runtime/panic.go:679 +0x1b2
sigs.k8s.io/k8s-container-image-promoter/lib/dockerregistry.checkOverlappingEdges(0xc0001ef390, 0xc0001ef390, 0xc0001ef978, 0xc0001efad0)
	lib/dockerregistry/inventory.go:425 +0x226
sigs.k8s.io/k8s-container-image-promoter/lib/dockerregistry.ToPromotionEdges(0xc0003eb930, 0x2, 0x2, 0xc0000285a0, 0x60, 0xc0000285a0)
	lib/dockerregistry/inventory.go:283 +0x137
sigs.k8s.io/k8s-container-image-promoter/lib/dockerregistry.TestValidateManifestsFromDir(0xc000101200)
	lib/dockerregistry/inventory_test.go:677 +0x3d5
testing.tRunner(0xc000101200, 0x934600)
	GOROOT/src/testing/testing.go:909 +0xc9
created by testing.(*T).Run
	GOROOT/src/testing/testing.go:960 +0x350
```

I find this surprising. I guess I can't use the 2nd (DRY-er) form...?